session.go

```diff
@@ -31,182 +32,131 @@ const (
-		options:    options,
-		state:      StateStarting,
-		socketsDir: filepath.Join(tempdir, "sockets"),
+		attachTimeout: 30 * time.Second,
+		command:       command,
+		cond:          sync.NewCond(&sync.Mutex{}),
+		configFile:    filepath.Join(tempdir, "config"),
+		execer:        execer,
+		id:            uuid.NewString(),
+		options:       options,
+		state:         StateStarting,
+		socketsDir:    filepath.Join(tempdir, "sockets"),
 	}
-	go s.lifecycle(id)
+	go s.lifecycle()
 	return s
 }
 
 // lifecycle manages the lifecycle of the session.
-func (s *Session) lifecycle(id string) {
-	// When this context is done the process is dead.
-	ctx, cancel := context.WithCancel(context.Background())
-
-	// Close the session down immediately if there was an error.
-	process, err := s.start(ctx, id)
+func (s *Session) lifecycle() {
+	err := s.ensureSettings()
 	if err != nil {
-		defer cancel()
-		s.setState(StateDone, xerrors.Errorf("process start: %w", err))
+		s.setState(StateDone, xerrors.Errorf("ensure settings: %w", err))
 		return
 	}
 
-	// Close the session after a timeout.  Timeouts created with AfterFunc can be
-	// reset; we will reset it as long as there are active connections.  The
-	// initial timeout for starting up is set here and will probably be less than
-	// the session timeout in most cases.  It should be at least long enough for
-	// screen to be able to start up.
-	s.timer = time.AfterFunc(30*time.Second, s.Close)
+	// The initial timeout for starting up is set here and will probably be far
+	// shorter than the session timeout in most cases.  It should be at least long
+	// enough for the first screen attach to be able to start up the daemon.
+	s.timer = time.AfterFunc(s.attachTimeout, s.Close)
 
-	// Emit the done event when the process exits.
-	go func() {
-		defer cancel()
-		err := process.Wait()
-		if err != nil {
-			err = xerrors.Errorf("process exit: %w", err)
-		}
-		s.setState(StateDone, err)
-	}()
-
-	// Handle the close event by killing the process.
+	// Handle the close event by asking screen to quit the session.  We have no
+	// way of knowing when the daemon process dies so the Go side will not get
+	// cleaned up until the timeout if the process gets killed externally (for
+	// example via `exit`).
 	go func() {
 		s.waitForState(StateClosing)
 		s.timer.Stop()
-		// process.Close() will send a SIGTERM allowing screen to clean up.  If we
-		// kill it abruptly the socket will be left behind which is probably
-		// harmless but it causes screen -list to show a bunch of dead sessions.
-		// The user would likely not see this since we have a custom socket
-		// directory but it seems ideal to let screen clean up anyway.
-		process.Close()
-		select {
-		case <-ctx.Done():
-			return // Process exited on its own.
-		case <-time.After(5 * time.Second):
-			// Still running; cancel the context to forcefully terminate the process.
-			cancel()
-		}
+		err := s.sendCommand(context.Background(), "quit")
```

Comment:
I have a slight worry about leaking this goroutine if our quit command has a hard failure we do not forsee.  Like, let's say that some new or obscure version of screen has different flags.

Similarly, I worry about leaking screen sessions if `execer.Start()` throws an error.

Do we need to think about a GC-based architecture?

Or at least some logging so we have a fighting chance of noticing problems like the above?

Reply:
I definitely worry about the same.  100% agree with adding a log at the very least.

I was considering some garbage collection based off `screen -list` but I think that has a similar problem of possibly changing in terms of flags or output.

What about a goroutine that iterates through `/proc` for `screen` processes that match on UUIDs that have not yet been confirmed to be killed then kills them through their PID if they are still running?

Reply:
Let's start with logs and make a point to keep an eye on it.  `screen` has been around for absolutely ages, so maybe my paranoia isn't justified.

Reply:
Sounds good!